[clang-cpp] class inheritance

kunjsong01 opened the pull request:
Add support for inheritance and polymorphism:

kunjsong01 committed:
[show_symbol] added debug lines

kunjsong01 committed:
[clang-cpp] added base method parsing

kunjsong01 committed:
[clang-cpp] added base init parsing

kunjsong01 commented:
@mikhailramalho 
I'm currently researching the best way to implement vtable but not sure if we should follow approach in old frontend (cpp_typecheck_virtual_table.cpp).

Here's what I found in clang: 
Clang appears to generate vtable in CodeGen: 
https://github.com/llvm/llvm-project/blob/release/11.x/clang/lib/CodeGen/CGVTables.h
https://clang.llvm.org/doxygen/classclang_1_1CodeGen_1_1CodeGenVTables.html

I reckon it would save us a lot time and effort if we can use clang's infrastructure. But here comes the questions: 
 
Q1. Is it to use CodeGen to generate a vtable along with the AST? 
Q2. Are we able to use the in-memory form of Clang's vtable in the same way as we use the in-memory form of Clang AST for typechecking?

lucasccordeiro reviewed (changes requested):
@kunjsong01: Can I ask you whether you could add some regression tests or unit tests?

kunjsong01 committed:
[regression] enabled test case for inheritance

kunjsong01 commented:
> @kunjsong01: Can I ask you whether you could add some regression tests or unit tests?

Done. Enabled two test cases in commit d10f6d13f60ea3f437bfbbd0b8837b6049e3a948 including pass and fail scenarios.

kunjsong01 committed:
[regression] added cpp-inheritance test suite

mikhailramalho commented:
> @mikhailramalho I'm currently researching the best way to implement vtable but not sure if we should follow approach in old frontend (cpp_typecheck_virtual_table.cpp).
> 
> Here's what I found in clang: Clang appears to generate vtable in CodeGen: https://github.com/llvm/llvm-project/blob/release/11.x/clang/lib/CodeGen/CGVTables.h https://clang.llvm.org/doxygen/classclang_1_1CodeGen_1_1CodeGenVTables.html
> 
> I reckon it would save us a lot time and effort if we can use clang's infrastructure. But here comes the questions:
> 
> Q1. Is it to use CodeGen to generate a vtable along with the AST? 
> Q2. Are we able to use the in-memory form of Clang's vtable in the same way as we use the in-memory form of Clang AST for typechecking?

I don't know :)

Using clang to generate the vtable seems like a better way forward and should play nice with our "convert then adjust" method, however, I can't help you with that as I never touched the codegen codebase.

kunjsong01 commented:
>Using clang to generate the vtable seems like a better way forward and should play nice with our "convert then adjust" method, however, I can't help you with that as I never touched the codegen codebase.

Thanks for sharing your thoughts! I'm going to implement the same method on the old frontend to build the virtual table, since it has been proved to work before. 
As for the clang vtable, I feel there is still a lot of things to learn and experiment with. If I can find an elegant way to generate the vtable, we can refactor the code for vtable generation.

fbrausse commented:
@kunjsong01 Do I understand correctly that you want to implement the same approach taken by compilers to model virtual methods? I.e. each non-abstract class gets a `vptr` member which on construction is initialized to point to another object (the type's vtable) containing the relevant function pointers?

I wonder whether instead of `vptr` it might be more efficient to store a type identifier based on which the correct method to call would be chosen by ESBMC. This would avoid the double indirection of having the vtable as another struct storing function pointers.

Please ignore if this idea is not feasible due to the current implementation of dynamic dispatch, I haven't checked how ESBMC is doing it now.

kunjsong01 commented:
> @kunjsong01 Do I understand correctly that you want to implement the same approach taken by compilers to model virtual methods? I.e. each non-abstract class gets a `vptr` member which on construction is initialized to point to another object (the type's vtable) containing the relevant function pointers?
> 

Yes!  

>I wonder whether instead of vptr it might be more efficient to store a type identifier based on which the correct method to call would be chosen by ESBMC. This would avoid the double indirection of having the vtable as another struct storing function pointers.
>
>Please ignore if this idea is not feasible due to the current implementation of dynamic dispatch, I haven't checked how ESBMC is doing it now.

I managed to generate the GOTO functions for test case `regression/esbmc-cpp/inheritance/paper_inheritance_example`. It appears that ESBMC adds additional GOTO statement(s) to set the `vptr` member in the constructor and the dynamic dispatch relies on such `vptr`.

kunjsong01 commented:
A note for future clang vtable approach: 
https://clang.llvm.org/doxygen/VTableBuilder_8h.html
https://clang.llvm.org/doxygen/CXXInheritance_8h.html 

`clang/AST/CXXInheritance.h` and `clang/AST/VTableBuilder.h` might be a good starting point. (For the time being, I'm not seeing the APIs to generate/use any in-memory form vtable. )

mikhailramalho commented:
the VTableBuilder looks interesting.

Em ter., 6 de set. de 2022 às 11:00, KSong ***@***.***>
escreveu:

> A note for future clang vtable approach:
> https://clang.llvm.org/doxygen/VTableBuilder_8h.html
> https://clang.llvm.org/doxygen/CXXInheritance_8h.html
>
> clang/AST/CXXInheritance.h and clang/AST/VTableBuilder.h might be a good
> starting point. (For the time being, I'm not see the APIs to generate/use
> any in-memory form vtable. )
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/848#issuecomment-1238193094>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH5P7KZS24DPHLABCQ3V45E7NANCNFSM6AAAAAAQEV2EOQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Mikhail Ramalho.

kunjsong01 committed:
[clang-c*-frontend] added debug functions

kunjsong01 commented:
TODOs to pass test case `paper_inheritance_example/main.cpp`:
1. Match symbols for Vechicle class (WIP)
2. Match symbols for Car class (TODO)
3. Match symbols for Motorcycle class (TODO)

```cpp
// Vehicle class
class Vehicle
{
public:
  Vehicle() {}
  virtual int number_of_wheels() = 0;
};
```

### Set of symbols generated by the old frontend for Vehicle class:
```python
      1 tag.Vehicle                     ---> class symbol
      1 Vehicle::number_of_wheels(this) ---> method symbol
      1 Vehicle::operator=(this,ref_symbol(identifier=tag.Vehicle,#constant=1_1)) ---> copy assignment optr symbol
      1 Vehicle::Vehicle(this,ref_symbol(identifier=tag.Vehicle,#constant=1_1))   ---> copy ctor symbol
      1 Vehicle::~Vehicle(this)  ---> dtor symbol
      1 Vehicle::Vehicle(this)   ---> ctor symbol
      1 Vehicle::~Vehicle(this)::Vehicle::~Vehicle(this)::this ---> dtor _this_ argument
      1 Vehicle::Vehicle(this)::Vehicle::Vehicle(this)::this   ---> ctor _this_ argument
      1 virtual_table::tag.Vehicle               ---> vtable symbol_1
      1 virtual_table::tag.Vehicle@tag.Vehicle   ---> vtable symbol_2
```

### Set of symbols generated by the clang-cpp-frontend for Vehicle class:
```python
      1 tag-Vehicle ---> class symbol
      1 c:@S@Vehicle@F@number_of_wheels# ---> method symbol
      1 c:@S@Vehicle@F@~Vehicle# ---> dtor symbol
      1 c:@S@Vehicle@F@Vehicle# ---> ctor symbol
      1 c:@S@Vehicle@F@Vehicle#&1$@S@Vehicle# ---> copy ctor
      1 c:@S@Vehicle@F@Vehicle#&1$@S@Vehicle#this ---> copy ctor _this_ argument 
      1 c:@S@Vehicle@F@~Vehicle#this ---> dtor _this_ argument
      1 c:@S@Vehicle@F@Vehicle#this ---> ctor _this_ argument
```

### Clang-cpp frontend extensions:
We should be able to generate the same GOTO functions by taking the union of the two sets of symbols. 
- Must have: {vtable_symbol_1, vtable_symbol_2}
- TBD: {copy assignment optr} 

### Findings:
1. The old parse tree does NOT have nodes for dtor, copy assignment optr or copy ctor. But the old frontend secretly conforms to the Rule of Three by adding symbols for those functions. 
2. Although the old frontend secretly adds copy assignment optr and copy ctor, the `this` argument symbols for these functions are missing in the symbol table.  
3. Vtable is added as a componenet to the class

kunjsong01 committed:
[cpp] added debugging functions

  - print intermediate symbol table

kunjsong01 committed:
[context] added debugging lines to catch target symbol

kunjsong01 committed:
[regression] temporarily removed cpp/inheritance test suite

kunjsong01 committed:
[cpp][util] various debugging lines to keep the old data

kunjsong01 committed:
[clang-c] added #virtual_name and #is_virtual

kunjsong01 committed:
[clang-c*] added virtual_table::tag-<class_name>

kunjsong01 committed:
[clang-c*] added virtual_table::tag-<class_name> - Part II

  - moved get_virtual_method to clang_cpp_converter (done)
  - replaced current_class_id and current_class_name with
  current_class_symbol (WIP)

kunjsong01 committed:
[clang-c*] added virtual_table::tag-<class_name> - Part III

  - matched old frontend symbol "virtual_table::tag.Vehicle"

kunjsong01 committed:
[clang-c*] replaced static_cast with llvm's checking cast

kunjsong01 committed:
[clang-c*] match symbol level - Part I

  - fixed clang headers
  - added components in s.type for tag-Vehicle symbol
  - skipped first implicit CXXRecordDecl note in class AST

kunjsong01 committed:
[clang-c*] fixed CI

kunjsong01 committed:
[clang-c*] fixed CI for Part I

kunjsong01 committed:
[clang-c*] match symbol level - Part II

   - skipped implicit stuff in class declarators
   - moved C++ virtual method processing to clang-cpp
   - flaged ctor and dtor

kunjsong01 committed:
[clang-c*] linting

kunjsong01 committed:
[clang-c*] fixed CI for Part II

kunjsong01 commented:
Closed and replaced by https://github.com/esbmc/esbmc/pull/854 due to too many experiment stuff.
